zilliqa/src/consensus.rs

```diff
@@ -1211,43 +1201,72 @@ impl Consensus {
             }
 
             // Apply specific txn
-            let result = Self::apply_transaction_at(
-                state,
-                self.db.clone(),
-                tx.clone(),
-                executed_block_header,
-                inspector::noop(),
-            )?;
-
-            // Skip transactions whose execution resulted in an error and drop them.
-            let Some(result) = result else {
-                continue;
+            // Depending on the outcome of apply_transaction:
+            // 1. Transaction is successful but cannot fit into block due to gas limit - return it back to mempool
+            // 2. Transaction is successful and fits into block
+            // 3. Transaction failed and cannot fit into block due to gas limit - return it back to mempool
+            // 4. Transaction failed but can be included in a block (marked as failed). There's cost incurred.
+            let (gas_exceeded, receipt) = {
+                match Self::apply_transaction_at(
+                    state,
+                    self.db.clone(),
+                    tx.clone(),
+                    executed_block_header,
+                    inspector::noop(),
+                ) {
+                    Ok(result) => {
+                        if let Some(g) = gas_left.checked_sub(result.gas_used()) {
+                            gas_left = g;
+                            (
+                                false,
+                                Some(Self::create_txn_receipt(
+                                    result.clone(),
+                                    tx.hash,
+                                    tx_index_in_block,
+                                    self.config.consensus.eth_block_gas_limit - gas_left,
+                                )),
+                            )
+                        } else {
+                            (true, None)
+                        }
+                    }
+                    Err(_) => {
```

Comment:
I'm not convinced this branch makes sense. If `apply_transaction_at` fails for some catastrophic reason like the DB being inaccessible, then I don't think it makes sense to add add a failed transaction receipt - We should just propagate the error up and eventually panic.

If application-level errors (e.g. not enough funds) and system-level errors are both getting muddled into the `Err` half of the `Result` of `apply_transaction_at` then I think we should fix that lower down the stack. Or in other words, `State::apply_transaction` should always return `Ok(TransactionApplyResult)` if we expect it to produce a mined transaction and a receipt.

In theory I think we could make this change by matching on the result of `evm.transact()` and mapping some of the `EvmError` variants to `Ok(<failed transaction receipt>)` and keep others as an `Err`?